### PR TITLE
Reader: Add in-stream post recommendations

### LINKED ISSUE #9194: Reader: Implement in-stream recs
Starting with just post recommendations. See "In-Stream Recommendations Status Update" on the Readersquad p2 for details.

dismiss-2-wide

### PR DESCRIPTION
**Pull request by blowery**
This adds in-stream recommendations to the refreshed reader following stream. 

Things to test:
- [x] Keyboard shortcuts. Next (`j`) and Previous (`k`) should skip recommendation blocks.
- [x] The distance between recs should increase as you subscribe to more things. It grows quickly at first, but slows down as you add more subs. Yay math.
- [x] Test all the streams, both in the refresh and out of it. This is a pretty major refactor of how the post lifecycle works. 

Fixes #9194

### CODE REVIEW

**Comment by blowery** [comment]
Will fix tests in a bit.

File: client/blocks/reader-related-card-v2/index.jsx [hunk 1]
```diff
@@ -27,9 +26,9 @@ function FeaturedImage( { image, href } ) {
 		} } ></div> );
 }
 
-function AuthorAndSiteFollow( { post, site } ) {
+function AuthorAndSiteFollow( { post } ) {
 	const siteUrl = getStreamUrl( post.feed_ID, post.site_ID );
-	const authorAndSiteAreDifferent = site.title.toLowerCase() !== post.author.name.toLowerCase();
+	const authorAndSiteAreDifferent = post.site_name.toLowerCase() !== post.author.name.toLowerCase();
```
Comment (samouri) [inline client/blocks/reader-related-card-v2/index.jsx]: awesome

File: client/lib/feed-stream-store/feed-stream.js [hunk 2]
```diff
@@ -175,32 +175,44 @@ assign( FeedStream.prototype, {
...
-				this.selectedIndex = nextIndex;
-				this.emit( 'change' );
-			} else {
-				this.selectNextItem( nextIndex );
-			}
+	selectNextItem: function( ) {
+		if ( this.selectedIndex == null ) {
+			return;
+		}
+		let nextIndex = this.selectedIndex + 1;
```
Comment (samouri) [inline client/lib/feed-stream-store/feed-stream.js]: I think you could simplify the this function by using lodash `findIndex`.
```javascript 
selectNextItem: function() {
    const nextIndex = findIndex( this.postKeys, this._isValidPostOrGap, this.selectedIndex );
    if ( this.nextIndex !== this.selectedIndex ) {
        this.selectedIndex = nextIndex;
        this.emitChange();
    }
}

```

File: client/lib/feed-stream-store/feed-stream.js [hunk 3]
```diff
@@ -175,32 +175,44 @@ assign( FeedStream.prototype, {
...
-		var nextIndex = selectedIndex - 1;
-		if ( nextIndex > -1 && nextIndex < this.postKeys.length ) {
-			if ( this._isValidPostOrGap( this.postKeys[ nextIndex ] ) ) {
-				this.selectedIndex = nextIndex;
-				this.emit( 'change' );
-			} else {
-				this.selectPrevItem( nextIndex );
-			}
+	selectPrevItem: function() {
+		if ( ! this.selectedIndex ) {
```
Comment (samouri) [inline client/lib/feed-stream-store/feed-stream.js]: same comments above but for `findLastIndex`
```javascript 
const nextIndex = findLastIndex( this.postKeys, this._isValidPostOrGap, this.selectedIndex );
```

File: client/reader/stream/index.jsx [hunk 4]
```diff
@@ -60,10 +61,56 @@ function refreshCardFactory( post ) {
...
+const MAX_DISTANCE_BETWEEN_RECS = 49; // page size is 7, so 7 pages
+const RECS_PER_BLOCK = 2;
+
+function getDistanceBetweenRecs() {
+	// the distance between recs changes based on how many subscriptions the user has.
+	// We cap it at MAX_DISTANCE_BETWEEN_RECS.
+	// It grows at the natural log of the number of subs, times a multiplier, offset by a constant.
+	// This lets the distance between recs grow quickly as you add subs early on, and slow down as you
+	// become a common user of the reader.
+	const totalSubs = FeedSubscriptionStore.getTotalSubscriptions();
```
Comment (samouri) [inline client/reader/stream/index.jsx]: heres what it looks like on a [wolfram alpha graph](http://tinyurl.com/h3gamtq
)

File: client/blocks/reader-related-card-v2/index.jsx (hunk 1 above)
Comment (blowery) [inline client/blocks/reader-related-card-v2/index.jsx]: yeah, though `title` is ... fancy. If the site is missing an actual title, it falls back to the domain. We lost that with this change. Hrm.

File: client/reader/stream/index.jsx (hunk 4 above)
Comment (samouri) [inline client/reader/stream/index.jsx]: currently:
- from 1-9: is equals 4
- from 10-19 it increases 5-->8
- from 20-30 it increases from 9-->11
- ....
- from 400-600 it increases from 24-->26
- etc.

I don't feel super strong about this, but I'm not sure what the benefit to the user is of having this work so continuously as opposed to as a step function.  aka:
```javascript
if ( subs < 10 )
	return 4;
} else if ( subs < 20) {
	return 8;
else if ( subs < 50 ) {
	return 12;
} else if ( subs < 100) {
	return 20
} else if ( subs < 1000) {
	return 30;
}
```

The only reason I prefer a step function is that it is easier to reason about. If you think the UX of the continuous version is better, then lets stick with it 👍

File: client/reader/stream/index.jsx [hunk 5]
```diff
@@ -365,9 +449,13 @@ export default class ReaderStream extends React.Component {
 	}
 
 	renderPost = ( postKey, index ) => {
-		let postState,
-			PostClass;
-		const isSelected = index === this.state.selectedIndex;
+		const selectedPostKey = this.props.store.getSelectedPost();
+		const isSelected = selectedPostKey &&
+			selectedPostKey.postId === postKey.postId &&
+			(
```
Comment (blowery) [inline client/reader/stream/index.jsx]: we have two different sources of posts. posts from a feed have a feed id, posts from a site have a blog id. so anything from /read/following or /read/feeds/:feed/* will have a feed id, while anything from /read/blogs/* or /read/tag/* or search will generally have a blog id.

File: client/reader/stream/index.jsx [hunk 6]
```diff
@@ -60,10 +61,56 @@ function refreshCardFactory( post ) {
...
+
+	const itemsBetweenRecs = getDistanceBetweenRecs();
+	// if we don't have enough posts to insert recs, don't bother
+	if ( posts.length < itemsBetweenRecs ) {
+		return posts;
+	}
+
+	const items = [];
+
+	for ( let postIndex = 0, recIndex = 0; postIndex < posts.length; postIndex += itemsBetweenRecs, recIndex += RECS_PER_BLOCK ) {
```
Comment (blowery) [inline client/reader/stream/index.jsx]: This gets called when either the posts or the recs array changes.

Yeah, abusing push like this probably isn't the best method. I'm trying to do something that will allocate as few new arrays as possible, so I was trying to avoid using `concat`, which always makes a new array. 

`splice` might be the right way to go here?

**Comment by jancavan** [comment]
I've fixed some alignment and padding issues in: https://github.com/Automattic/wp-calypso/pull/9490/commits/3a8a84879ea86fb0166b81277d8b9689c07708ac:

screenshot-2016-11-23-15 38 24

But there's still some display issues in certain viewports that I'm trying to figure out due to the custom breakpoints we have:

screenshot 2016-11-23 16 47 18

File: client/reader/stream/index.jsx (hunk 4 above)
Comment (blowery) [inline client/reader/stream/index.jsx]: Tried this and it just ends up being the math blown out. I'd like to stick with the math for now.

File: client/reader/stream/index.jsx (hunk 6 above)
Comment (blowery) [inline client/reader/stream/index.jsx]: tried a new approach using `flatMap`.

**Comment by blowery** [comment]
Thanks @jan!

File: client/reader/stream/index.jsx [hunk 7]
```diff
@@ -91,19 +91,22 @@ function injectRecommendations( posts, recs = [] ) {
...
 
-	const items = [];
+	let recIndex = 0;
 
-	for ( let postIndex = 0, recIndex = 0; postIndex < posts.length; postIndex += itemsBetweenRecs, recIndex += RECS_PER_BLOCK ) {
-		items.push.apply( items, posts.slice( postIndex, postIndex + itemsBetweenRecs ) );
-		// only insert recs if we have them and if we filled a full block of posts
-		if ( recIndex < recs.length && postIndex + itemsBetweenRecs < posts.length ) {
-			items.push( {
+	return flatMap( posts, ( post, index ) => {
```
Comment (samouri) [inline client/reader/stream/index.jsx]: nice!

File: client/lib/feed-stream-store/feed-stream.js [hunk 8]
```diff
@@ -4,6 +4,8 @@
 var assign = require( 'lodash/assign' ),
```
Comment (bluefuton) [inline client/lib/feed-stream-store/feed-stream.js]: Worth cleaning up the imports while we're in there?

**Comment by samouri** [comment]
`injectRecommendations()` was changed in a previous [commit](https://github.com/Automattic/wp-calypso/pull/9490/commits/f85c7b5542aee6d7760460c2cacc9a4c071a2917) to  depend on the modulus of `distanceBetweenRecs`.  I believe this didn't work because oftentimes that number wasn't an Integer.  Pushed a fix (Math.floor) in [5c90300](https://github.com/Automattic/wp-calypso/pull/9490/commits/5c903009b5ce189346f2b8b0bbf68cf752315880)

File: client/reader/stream/post-lifecycle.jsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		}
+
+		switch ( postState ) {
+			case 'pending':
+				return <PostPlaceholder />;
+			case 'error':
+				return <PostUnavailable post={ post } />;
+			default:
+				const PostClass = this.props.cardClassForPost( post );
+				if ( PostClass === CrossPost ) {
```
Comment (samouri) [inline client/reader/stream/post-lifecycle.jsx]: can we remove this if clause?
```javascript
default:
  const PostClass = ...
  const xMetadata = XPostHelper.get...  // returns object full of nulls if not an xpost
  return <PostClass 
                  ....all the fields
                />
```

File: client/reader/stream/index.jsx [hunk 10]
```diff
@@ -196,11 +266,18 @@ export default class ReaderStream extends React.Component {
 			if ( postKey.isGap === true ) {
 				return this._selectedGap.handleClick();
 			}
+
```
Comment (samouri) [inline client/reader/stream/index.jsx]: does `showSelectedPost()` now belong in `PostLifecyle` ?

File: client/reader/stream/post-lifecycle.jsx (hunk 9 above)
Comment (blowery) [inline client/reader/stream/post-lifecycle.jsx]: likely, but let's save that for another PR

File: client/lib/feed-stream-store/feed-stream.js (hunk 8 above)
Comment (blowery) [inline client/lib/feed-stream-store/feed-stream.js]: I'd rather do that in a separate PR to move all of feed stream over to es6

File: client/reader/stream/index.jsx [hunk 11]
```diff
@@ -26,14 +25,16 @@ import Post from './post';
...
 import page from 'page';
-import PostUnavailable from './post-unavailable';
 import PostPlaceholder from './post-placeholder';
 import PostStore from 'lib/feed-post-store';
 import UpdateNotice from 'reader/update-notice';
 import PostBlocked from './post-blocked';
 import KeyboardShortcuts from 'lib/keyboard-shortcuts';
 import scrollTo from 'lib/scroll-to';
 import XPostHelper from 'reader/xpost-helper';
+import RecommendationBlock from './recommendation-block';
```
Comment (blowery) [inline client/reader/stream/index.jsx]: done

File: client/reader/stream/index.jsx (hunk 10 above)
Comment (blowery) [inline client/reader/stream/index.jsx]: don't think so, that would mean moving the keyboard event binding there too.

File: client/reader/stream/index.jsx (hunk 5 above)
Comment (blowery) [inline client/reader/stream/index.jsx]: no, should be two different post key instances. that said, there's no guarantee that there's only one post key instance for a given id pair. We can't just use `===` to do the compare.

a post key has _either_ a blogId or a feedId, but never both. In retrospect, we probably should have used a different property for each site for the post id. But, future PR as that will touch a ton of stuff.

**Comment by blowery** [comment]
addressed all the stuff from review

**Comment by blowery** [comment]
added in stats, changed to defaulting to max distance when we don't know how many subs a person has.

**Comment by samouri** [comment]
This is so close!
Recommended posts are not showing their 3 lines of text example

I believe it is this line of [code](https://github.com/Automattic/wp-calypso/pull/9490/files#diff-5fc366969c3080c52fb50121347ef0e3R467) that is responsible.  Was this intentional?

Also, should we merge this in before fixing all of the breakpoints?

**Comment by jancavan** [comment]
@samouri Yes, the two recs that are side-by-side are supposed to have the same height.

**Comment by blowery** [comment]
@jancavan i jumped up the min height to 250px to get the lines of text

**Comment by blowery** [comment]
Fixes #9194

**Comment by bluefuton** [comment]
Small proptype nit:

screen shot 2016-11-29 at 16 17 20

Otherwise looks good to me! 🚀
